dune: init at 2.1.0

marsam opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

marsam committed:
ocamlPackages.cppo: 1.6.5 -> 1.6.6

marsam committed:
ocamlPackages.dtoa: 0.3.1 -> 0.3.2

marsam committed:
ocamlPackages.wtf8: 1.0.1 -> 1.0.2

marsam committed:
ocamlPackages.ppx_gen_rec: 1.0.0 -> 1.1.0

marsam committed:
ocamlPackages.biniou: 1.2.0 -> 1.2.1

marsam commented:
@GrahamcOfBorg fstar

marsam commented:
@GrahamcOfBorg build fstar

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/explicit-runhook-in-derivation/4614/34

vyorkin commented:
Do you want to add `dune-configurator` as well or it would be better to add it in a separate PR?

Zimmi48 commented on pkgs/development/tools/ocaml/cppo/default.nix:
```diff
@@ -7,11 +7,15 @@ assert stdenv.lib.versionAtLeast ocaml.version "3.12";
+        runHook preInstall
+        ${dune}/bin/dune install --prefix=$out --libdir=$OCAMLFIND_DESTDIR
+        runHook postInstall
+      '';
```
I don't understand these changes. What are they meant to fix? If running `dune install` is generally better than running `opaline`, why not change this in the Dune derivation itself?

marsam replied:
I simply prefer to use dune install over opaline. I recall trying to set it to into `buildDunePackage` but there were some failing packages, but I haven't investigated much. 
I'll try to set into `buildDunePackage` it in another PR

Zimmi48 reviewed (commented):
Not having Dune 2.0.0 yet in nixpkgs is an inconvenience. Can you please explain what are the things that are blocking this PR or what additional work you think is required?

I am not sure I understand the first commit, adding Dune 2.0, but keeping Dune 1.11 without exposing it. Why would we want to keep Dune 1.11 in nixpkgs?

If this is because Dune 2.0 is not compatible with OCaml < 4.06, note that we could go back to the situation preceding #57953 where Dune was out of `ocamlPackages`. This was changed because Dune is both a tool and a library, but we could try to find another solution, e.g. making it available as a library only for `ocamlPackages` sets starting with 4.06 and as a tool independent of `ocamlPackages` for any OCaml version.

I suppose it is not because some OCaml packages have not been migrated yet from jbuilder to Dune, otherwise you wouldn't be updating a bunch of packages in this PR as well. BTW, for packages that still haven't migrated to Dune in their latest version, maybe that says something about them (not) being maintained, and we could simply choose to drop them, or at least mark them as broken and not care much...

marsam committed:
ocamlPackages.stdint: 0.5.1 -> 0.6.0

marsam committed:
ocamlPackages.rope: 0.6.1 -> 0.6.2

marsam committed:
ocamlPackages.ocp-index: use buildDunePackage

marsam committed:
ocamlPackages.utop: use buildDunePackage

marsam committed:
dune_2: init at 2.1.0

Changelog:
- https://github.com/ocaml/dune/releases/tag/2.0.0
- https://github.com/ocaml/dune/releases/tag/2.0.1
- https://github.com/ocaml/dune/releases/tag/2.1.0

marsam committed:
ocamlPackages.dune-private-libs: init at 2.1.0

marsam committed:
ocamlPackages.dune-configurator: init at 2.1.0

marsam commented on pkgs/development/ocaml-modules/stdint/default.nix:
```diff
@@ -13,8 +13,20 @@ stdenv.mkDerivation rec {
   buildInputs = [ ocaml findlib dune ];
 
+  patches = [
+    # Fetch dune patch @mor1 fork. TODO(marsam): create a PR in upstream
```
Thanks, fixed

marsam commented:
>Why would we want to keep Dune 1.11 in nixpkgs?

dune 2 is a major version bump, a there are multiple packages which still haven't been migrated, for instance, janestreet packages.

I tried to set dune 2 as default, but I encountered various errors, hence only added dune_2 only as another attribute. 

Only the last three commits are relevant; @vbgl should I cut the other commits into a separated PR?

Zimmi48 commented:
> > 
> 
> dune 2 is a major version bump, a there are multiple packages which still haven't been migrated, for instance, janestreet packages.

In theory, Dune 2 shouldn't break Dune projects which set the Dune language version to a lower version than 2.0. However, it does break the projects that have not migrated to Dune from jbuilder. Did you find examples beyond this? If yes, it could be interesting to look into them more closely and possibly report the issues to Dune developers.

In any case, this PR LGTM (in full, including the version bumps) and I'm in favor of merging it as-is now (without splitting).

Zimmi48 commented:
@marsam I'd like to encourage you to go ahead and merge this without waiting further.

Zimmi48 commented:
@marsam or @vbgl: please merge this now.

BTW: PR title should be edited to reflect the change (dune_2: init at 2.1.0).

marsam commented:
@Zimmi48 sorry for the delay, I was waiting for another review; merging as-is because it shouldn't introduce any major change. @vbgl I hope you don't mind

Zimmi48 commented:
Thanks!

Shados commented:
@marsam This PR included an update for cppo from 1.6.5 to 1.6.6. This broke building cppo with ocaml 4.02, which is still needed for some things.

marsam commented:
@Shados sorry about that, reverted in https://github.com/NixOS/nixpkgs/pull/76942
